File: erlang/3.4/geom.erl
%%% -----------------------------------------------
%%% @author anders pearson
%%% @doc simple geometry module
%%% @end
%%% -----------------------------------------------
-module(geom).
-export([area/1]).

%% @spec area({T::Atom, X::integer(), Y::integer()}) -> integer()
%% @doc calculates area of the given shape.
area({Type, A, B}) ->
		area(Type, A, B).

%% @spec area(Atom, X::integer(), Y::integer()) -> integer()
%% @doc calculates area of the given shape.
area(rectangle, X, Y) when X > 0, Y > 0 ->
		X * Y;
area(triangle, X, Y) when X > 0, Y > 0 ->
		(X * Y) / 2;
area(ellipse, X, Y) when X > 0, Y > 0 ->
		math:pi() * X * Y;
area(_, _, _) ->
		0.

File: elixir/3.3/geom.ex
defmodule Geom do
	def area(:rectangle, width, height) when width >= 0 and height >= 0 do
		width * height
	end

  def area(:triangle, width, height) when width >= 0 and height >= 0 do
    width * height / 2.0
  end

  def area(:ellipse, width, height) when width >= 0 and height >= 0 do
    :math.pi() * width * height
  end

  def area(_, _, _) do
    0
  end
end
